File: app/src/main/java/com/ga/task/PriorityTaskProvider.java
package com.ga.task;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.SparseArray;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;

import static junit.framework.Assert.assertTrue;

/**
 * Created by alexeyglushkov on 30.12.14.
 */

// A task source for TaskManager

public class PriorityTaskProvider implements TaskProvider, TaskPool, Task.StatusListener {
    static final String TAG = "PriorityTaskProvider";

    private String providerId;
    private Handler handler;
    private Object userData;
    private List<TaskPoolListener> listeners;
    private int priority;

    // Task type -> priority queue
    private SparseArray<PriorityQueue<Task>> taskQueues;

    public PriorityTaskProvider(Handler handler, String id) {
        providerId = id;
        taskQueues = new SparseArray<PriorityQueue<Task>>();
        listeners = new ArrayList<TaskPoolListener>();
        setHandler(handler);
    }

    private PriorityQueue<Task> createQueueSet() {
        return new PriorityQueue<Task>(11, new Comparator<Task>() {
            @Override
            public int compare(Task lhs, Task rhs) {
                if (lhs == rhs) {
                    return 0;
                }

                if (lhs.getTaskPriority() > rhs.getTaskPriority()) {
                    return -1;
                }

                if (lhs.getTaskPriority() == rhs.getTaskPriority()) {
                    return 0;
                }

                return 1;
            }
        });
    }

    @Override
    public Task getTopTask(List<Integer> typesToFilter) {
        checkHandlerThread();

        Task topTask = null;
        int topPriority = -1;

        for (int i = 0; i < taskQueues.size(); i++) {
            if (typesToFilter == null || !typesToFilter.contains(taskQueues.keyAt(i))) {
                PriorityQueue<Task> queue = taskQueues.get(taskQueues.keyAt(i));
                Task queueTask = getTopTask(queue, false);

                if (queueTask != null && queueTask.getTaskPriority() > topPriority) {
                    topTask = queueTask;
                    topPriority = topTask.getTaskPriority();
                }
            }
        }

        return topTask;
    }

    @Override
    public Task takeTopTask(List<Integer> typesToFilter) {
        checkHandlerThread();

        Task task = getTopTask(typesToFilter);
        if (task != null) {
            PriorityQueue<Task> queue = taskQueues.get(task.getTaskType());
            Task polledTask = getTopTask(queue, true);
            triggerOnTaskRemoved(polledTask);
        }

        return task;
    }

    // TODO: find a better solution, TreeSet isn't great because it requires unique objects and
    // it's hard to achieve it with tasks using comparator, we just don't have enough information.
    // Also deletion in TreeSet happens via comparator and it must return 0 only for equal objects.
    private Task getTopTask(PriorityQueue<Task> queue, boolean needPoll) {
        Task topTask = null;

        ArrayList<Task> polledTasks = new ArrayList<Task>();
        while (!queue.isEmpty()) {
            if (!queue.peek().isBlocked()) {
                topTask = needPoll ? queue.poll() : queue.peek();
                break;
            } else {
                polledTasks.add(queue.poll());
            }
        }

        for (Task task : polledTasks) {
            queue.add(task);
        }

        return topTask;
    }

    public void updatePriorities(final PriorityProvider provider) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < taskQueues.size(); i++) {
                    PriorityQueue<Task> queue = createQueueSet();
                    for (Task t : taskQueues.get(taskQueues.keyAt(i))) {
                        t.setTaskPriority(provider.getPriority(t));
                        queue.add(t);
                    }

                    taskQueues.put(taskQueues.keyAt(i), queue);
                }
            }
        });
    }

    public void setTaskProviderId(String id) {
        providerId = id;
    }

    public String getTaskProviderId() {
        return providerId;
    }

    @Override
    public void setPriority(int priority) {
        this.priority = priority;
    }

    @Override
    public int getPriority() {
        return priority;
    }

    @Override
    public void addTask(final Task task) {
        if (!Tasks.isTaskReadyToStart(task)) {
            Log.d(TAG, "Can't put task " + task.getClass().toString() + " because it's already started " + task.getTaskStatus().toString());
            return;
        }

        // TaskProvider must set Waiting status on the current thread
        task.getPrivate().setTaskStatus(Task.Status.Waiting);

        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                addTaskOnThread(task);
            }
        });
    }

    @Override
    public void removeTask(final Task task) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                removeTaskOnThread(task);
            }
        });
    }

    public void onTaskStatusChanged(Task task, Task.Status oldStatus, Task.Status newStatus) {
        if (Tasks.isTaskCompleted(task)) {
            removeTask(task);
        }
    }

    @Override
    public void setHandler(Handler handler) {
        if (this.handler != null) {
            checkHandlerThread();
        }

        this.handler = handler;
    }

    @Override
    public Handler getHandler() {
        return handler;
    }

    private void addTaskOnThread(final Task task) {
        task.addTaskStatusListener(PriorityTaskProvider.this);
        addTaskToQueue(task);

        for (TaskPoolListener listener : listeners) {
            listener.onTaskAdded(PriorityTaskProvider.this, task);
        }
    }

    private void removeTaskOnThread(final Task task) {
        checkHandlerThread();

        if (removeTaskFromQueue(task)) {
            triggerOnTaskRemoved(task);
        }
    }

    private void triggerOnTaskRemoved(Task task) {
        checkHandlerThread();

        for (TaskPoolListener listener : listeners) {
            listener.onTaskRemoved(PriorityTaskProvider.this, task);
        }
    }

    private void addTaskToQueue(Task task) {
        checkHandlerThread();

        PriorityQueue<Task> queue = taskQueues.get(task.getTaskType());
        if (queue == null) {
            queue = createQueueSet();
            taskQueues.put(task.getTaskType(), queue);
        }

        queue.add(task);
    }

    private boolean removeTaskFromQueue(Task task) {
        checkHandlerThread();

        PriorityQueue<Task> queue = taskQueues.get(task.getTaskType());
        if (queue != null) {
            return queue.remove(task);
        }

        return false;
    }

    public interface PriorityProvider {
        int getPriority(Task task);
    }

    @Override
    public Task getTask(String taskId) {
        checkHandlerThread();

        Task resultTask = null;
        for (int i=0; i<taskQueues.size() && resultTask == null; ++i) {
            PriorityQueue<Task> queue = taskQueues.get(taskQueues.keyAt(i));
            Iterator<Task> iterator = queue.iterator();

            while (iterator.hasNext()) {
                Task task = iterator.next();
                if (task.getTaskId() != null && task.getTaskId().equals(taskId)) {
                    resultTask = task;
                    break;
                }
            }
        }

        return resultTask;
    }

    @Override
    public int getTaskCount() {
        checkHandlerThread();

        int resultCount = 0;
        for (int i=0; i<taskQueues.size(); ++i) {
            PriorityQueue<Task> queue = taskQueues.get(taskQueues.keyAt(i));
            resultCount += queue.size();
        }

        return resultCount;
    }

    @Override
    public List<Task> getTasks() {
        checkHandlerThread();

        ArrayList<Task> tasks = new ArrayList<Task>();

        for (int i=0; i<taskQueues.size(); ++i) {
            PriorityQueue<Task> queue = taskQueues.get(taskQueues.keyAt(i));
            tasks.addAll(queue);
        }

        return tasks;
    }

    @Override
    public void setUserData(Object data) {
        userData = data;
    }

    @Override
    public Object getUserData() {
        return userData;
    }

    @Override
    public void addListener(TaskPoolListener listener) {
        listeners.add(listener);
    }

    @Override
    public void removeListener(TaskPoolListener listener) {
        listeners.remove(listener);
    }

    private void checkHandlerThread() {
        Assert.assertEquals(Looper.myLooper(), handler.getLooper());
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/providers/PriorityTaskProvider.kt
package com.example.alexeyglushkov.taskmanager.providers

import androidx.annotation.WorkerThread

import androidx.collection.SparseArrayCompat
import com.example.alexeyglushkov.taskmanager.task.*
import com.example.alexeyglushkov.taskmanager.pool.TaskPoolBase
import com.example.alexeyglushkov.taskmanager.tools.SortedList
import com.example.alexeyglushkov.taskmanager.runners.ThreadRunner

import java.util.ArrayList
import java.util.Comparator

/**
 * Created by alexeyglushkov on 30.12.14.
 */

// A task source for TaskManager

open class PriorityTaskProvider(threadRunner: ThreadRunner, override var taskProviderId: String): TaskPoolBase(threadRunner), TaskProvider, Task.StatusListener {
    override var priority: Int = 0
    override var taskFilter: TaskProvider.TaskFilter? = null

    // Task type -> priority queue
    // TODO: try to use PriorityBlockingQueue
    private val taskQueues: SparseArrayCompat<SortedList<Task>> = SparseArrayCompat()

    override fun addTask(task: Task) {
        if (TaskProviders.addTaskCheck(task as TaskBase, tag())) {
            super.addTask(task)
        }
    }

    @WorkerThread
    override fun getTaskCount(): Int {
        checkHandlerThread()

        var resultCount = 0
        for (i in 0 until taskQueues.size()) {
            val queue = taskQueues.get(taskQueues.keyAt(i))
            resultCount += queue?.size ?: 0
        }

        return resultCount
    }

    @WorkerThread
    override fun getTasks(): List<Task> {
        checkHandlerThread()

        val tasks = ArrayList<Task>()
        for (i in 0 until taskQueues.size()) {
            val queue = taskQueues.get(taskQueues.keyAt(i))
            if (queue != null) {
                tasks.addAll(queue)
            }
        }

        return tasks
    }

    private fun createQueue(): SortedList<Task> {
        return SortedList(Comparator { lhs, rhs ->
            if (lhs.taskPriority == rhs.taskPriority) {
                return@Comparator 0
            }

            if (lhs.taskPriority > rhs.taskPriority) {
                -1
            } else 1
        })
    }

    @WorkerThread
    override fun getTopTask(): Task? {
        return getTopTask(false)
    }

    private fun getTopTask(needPoll: Boolean): Task? {
        checkHandlerThread()

        var topTask: Task? = null
        var topPriority = -1
        var topQueue: SortedList<Task>? = null
        val taskFilter = taskFilter

        for (i in 0 until taskQueues.size()) {
            if (taskFilter == null || !taskFilter.getFilteredTaskTypes().contains(taskQueues.keyAt(i))) {
                val queue = taskQueues.get(taskQueues.keyAt(i))
                if (queue != null) {
                    val queueTask = getTopTask(queue, false)
                    if (queueTask != null && queueTask.taskPriority > topPriority) {
                        topQueue = queue
                        topTask = queueTask
                        topPriority = topTask.taskPriority
                    }
                }
            }
        }

        if (needPoll && topQueue != null) {
            val resultTopTask = getTopTask(topQueue, true) // do poll
            assert(resultTopTask == topTask)
        }

        return topTask
    }

    @WorkerThread
    override fun takeTopTask(): Task? {
        checkHandlerThread()

        val task = getTopTask(true)
        if (task != null) {
            onTaskRemoved(task)
        }

        return task
    }

    @WorkerThread
    private fun getTopTask(queue: SortedList<Task>, needPoll: Boolean): Task? {
        var topTask: Task? = null
        val iterator = queue.iterator()
        while (iterator.hasNext()) {
            val aTask = iterator.next()
            val taskFilter = taskFilter
            val isFiltered = taskFilter != null && taskFilter.isFiltered(aTask)

            if (!isFiltered && !aTask.isBlocked()) {
                topTask = aTask
                if (needPoll) {
                    iterator.remove()
                }
                break
            }
        }

        return topTask
    }

    fun updatePriorities(provider: PriorityProvider) {
        threadRunner.launch {
            for (i in 0 until taskQueues.size()) {
                val queue = taskQueues.valueAt(i)
                val tasks = taskQueues.get(taskQueues.keyAt(i))
                if (tasks != null) {
                    for (t in tasks) {
                        t.taskPriority = provider.getPriority(t)
                    }
                }

                queue.updateSortedOrder()
            }
        }
    }

    @WorkerThread
    override fun addTaskInternal(task: Task) {
        addTaskToQueue(task)
    }

    @WorkerThread
    override fun removeTaskInternal(task: Task): Boolean {
        return removeTaskFromQueue(task)
    }

    @WorkerThread
    private fun addTaskToQueue(task: Task) {
        checkHandlerThread()

        var queue = taskQueues.get(task.taskType)
        if (queue == null) {
            queue = createQueue()
            taskQueues.put(task.taskType, queue)
        }

        queue.addInSortedOrder(task)
    }

    @WorkerThread
    private fun removeTaskFromQueue(task: Task): Boolean {
        checkHandlerThread()

        val queue = taskQueues.get(task.taskType)
        return queue?.remove(task) ?: false

    }

    interface PriorityProvider {
        fun getPriority(task: Task): Int
    }

    @WorkerThread
    override fun getTask(taskId: String): Task? {
        checkHandlerThread()

        var resultTask: Task? = null
        var i = 0
        while (i < taskQueues.size() && resultTask == null) {
            val queue = taskQueues.get(taskQueues.keyAt(i))
            queue?.let {
                val iterator = queue.iterator()
                while (iterator.hasNext()) {
                    val task = iterator.next()
                    if (task.taskId != null && task.taskId == taskId) {
                        resultTask = task
                        break
                    }
                }
            }

            ++i
        }

        return resultTask
    }

    override fun tag() = "PriorityTaskProvider"
}
